src/server/engine_shard_set.cc

```diff
@@ -56,6 +106,41 @@ EngineShard::Stats& EngineShard::Stats::operator+=(const EngineShard::Stats& o)
   return *this;
 }
 
+// for now this does nothing
+uint64_t EngineShard::DoDefragIfRequired(uint64_t cursor) {
+  // Placeholder for now!
+  return cursor;
+}
+
+// the memory de-fragmenting task is as follow:
+//  1. Check if memory usage is high enough
+//  2. Check if diff between commited and used memory is high enough
+//  3. Check if we have memory changes (to ensure that we not running endlessly).
+//  4. if all the above pass -> run on the shard and try to de-fragment memory by re-allocating
+//  values
+//     if the cursor for this is signal that we are not done, schedule the task to run at high
+//     priority otherwise lower the task priority so that it would not use the CPU when not required
+void EngineShard::StartDefragTask(util::ProactorBase* pb) {
+  uint64_t commited = GetMallocCurrentCommitted();
+  uint64_t mem_in_use = used_mem_current.load(memory_order_relaxed);
+
+  defrag_task_ = pb->AddOnIdleTask([this, commited, mem_in_use, cursor = 0ul]() mutable {
```

Comment:
I prefer to factor out this lambda into member function and keep its state (whatever you need) in member variables (could be a dedicated struct too).

Reply:
OK, just that this class would become larger..